File: 0049-group-anagrams/0049-group-anagrams.py
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # res = defaultdict(list)

        # for s in strs:
        #     count = [0] * 26

        #     for c in s:
        #         count[ord(c) - ord("a")] += 1
            
        #     res[tuple(count)].append(s)
            
        # return list(res.values())

        anagram_map = defaultdict(list)

        for word in strs:
            key = ''.join(sorted(word))
            anagram_map[key].append(word)
        
        return list(anagram_map.values())

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anagram = new HashMap<>();

        for (String str: strs){
            char[] chars = str.toCharArray();
            Arrays.sort(chars);
            String sorted = new String(chars);

            if(!anagram.containsKey(sorted)) {
                anagram.put(sorted, new ArrayList<>());
            }
            anagram.get(sorted).add(str);
        }
        return new ArrayList<>(anagram.values());
    }
}